File: internal/forms/federal/schedule_d.go
package federal

import (
	"taxpilot/internal/forms"
)

func init() { forms.RegisterForm(ScheduleD) }

// ScheduleD returns the FormDef for Schedule D — Capital Gains and Losses.
// This form aggregates capital gains/losses from:
//   - Form 8949 (stock/security sales from 1099-B)
//   - 1099-DIV Box 2a (capital gain distributions from mutual funds)
//
// Schedule D line 16 (net capital gain/loss) flows to Schedule 1 line 7,
// which then flows to Form 1040 line 7.
func ScheduleD() *forms.FormDef {
	return &forms.FormDef{
		ID:           forms.FormScheduleD,
		Name:         "Schedule D — Capital Gains and Losses",
		Jurisdiction: forms.Federal,
		TaxYears:      []int{2025},
		QuestionGroup: forms.GroupIncome1099,
		QuestionOrder: 3,
		Fields: []forms.FieldDef{
			// --- Part I: Short-Term Capital Gains and Losses ---

			// Line 1: Short-term from Form 8949 Box A (basis reported)
			// For MVP, all short-term goes here
			{
				Line:      "1",
				Type:      forms.Computed,
				Label:     "Short-term from Form 8949",
				DependsOn: []string{forms.F8949STGainLossKey},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F8949STGainLossKey)
				},
			},
			// Line 7: Net short-term capital gain or (loss)
			{
				Line:      "7",
				Type:      forms.Computed,
				Label:     "Net short-term capital gain or (loss)",
				DependsOn: []string{forms.SchedDLine1},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.SchedDLine1)
				},
			},

			// --- Part II: Long-Term Capital Gains and Losses ---

			// Line 8: Long-term from Form 8949 Box D (basis reported)
			// For MVP, all long-term goes here
			{
				Line:      "8",
				Type:      forms.Computed,
				Label:     "Long-term from Form 8949",
				DependsOn: []string{forms.F8949LTGainLossKey},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F8949LTGainLossKey)
				},
			},
			// Line 13: Capital gain distributions (from 1099-DIV Box 2a)
			{
				Line:      "13",
				Type:      forms.Computed,
				Label:     "Capital gain distributions",
				DependsOn: []string{forms.F1099DIVWildcardCapGain},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAll(forms.F1099DIVWildcardCapGain)
				},
			},
			// Line 15: Net long-term capital gain or (loss)
			{
				Line:      "15",
				Type:      forms.Computed,
				Label:     "Net long-term capital gain or (loss)",
				DependsOn: []string{forms.SchedDLine8, forms.SchedDLine13},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.SchedDLine8) + dv.Get(forms.SchedDLine13)
				},
			},

			// --- Part III: Summary ---

			// Line 16: Combined net gain or (loss)
			{
				Line:      "16",
				Type:      forms.Computed,
				Label:     "Net capital gain or (loss)",
				DependsOn: []string{forms.SchedDLine7, forms.SchedDLine15},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.SchedDLine7) + dv.Get(forms.SchedDLine15)
				},
			},
		},
	}
}

File: api/src/forms/federal/schedule_d.rs
use crate::domain::builders::*;
use crate::domain::field::*;
use crate::domain::form::*;

pub fn schedule_d() -> FormDef {
    FormDef {
        id: FORM_SCHEDULE_D.to_string(),
        name: "Schedule D — Capital Gains and Losses".to_string(),
        jurisdiction: Jurisdiction::Federal,
        tax_years: vec![2024, 2025, 2026],
        question_group: "income_1099".to_string(),
        question_order: 3,
        fields: vec![
            // --- Part I: Short-Term ---
            ref_field("1", "Short-term from Form 8949", F8949_ST_GAIN_LOSS),
            ref_field("7", "Net short-term capital gain or (loss)", SCHED_D_LINE_1),
            // --- Part II: Long-Term ---
            ref_field("8", "Long-term from Form 8949", F8949_LT_GAIN_LOSS),
            // Line 13: Capital gain distributions
            wildcard_sum_field("13", "Capital gain distributions", F1099_DIV_WILDCARD_CAP_GAIN),
            // Line 15
            sum_field("15", "Net long-term capital gain or (loss)", vec![SCHED_D_LINE_8, SCHED_D_LINE_13]),
            // --- Part III: Summary ---
            // Line 16
            sum_field("16", "Net capital gain or (loss)", vec![SCHED_D_LINE_7, SCHED_D_LINE_15]),
        ],
    }
}
